mm2src/kdf_walletconnect/src/lib.rs

```diff
@@ -0,0 +1,546 @@
+        //
+        // let wait_duration = Duration::from_secs(60);
+        // if let Ok(Some(resp)) = self.message_rx.lock().await.next().timeout(wait_duration).await {
+        //     let result = resp.mm_err(WalletConnectError::InternalError)?;
+        //     if let ResponseParamsSuccess::SessionEvent(data) = result.data {
+        //         if !data {
+        //             return MmError::err(WalletConnectError::PayloadError(
+        //                 "Please approve chain id change".to_owned(),
+        //             ));
+        //         }
+        //
+        //         self.session
+        //             .get_session_mut(&session.topic)
+        //             .ok_or(MmError::new(WalletConnectError::SessionError(
+        //                 "No active WalletConnect session found".to_string(),
+        //             )))?
+        //             .set_active_chain_id(chain_id.clone())
+        //             .await;
+        //     }
+        // }
+
+        Ok(())
+    }
+
+    /// TODO: accept WcChainId
+    /// Retrieves the available account for a given chain ID.
+    pub async fn get_account_for_chain_id(&self, chain_id: &WcChainId) -> MmResult<String, WalletConnectError> {
+        let namespaces = &self
+            .session
+            .get_session_active()
+            .await
+            .ok_or(MmError::new(WalletConnectError::SessionError(
+                "No active WalletConnect session found".to_string(),
+            )))?
+            .namespaces;
+
+        if let Some(Namespace {
+            accounts: Some(accounts),
+            ..
+        }) = namespaces.get(chain_id.chain.as_ref())
+        {
+            if let Some(account) = find_account_in_namespace(accounts, &chain_id.id) {
+                return Ok(account);
+            }
+        };
+
+        MmError::err(WalletConnectError::NoAccountFound(chain_id.to_string()))
+    }
+
+    /// Waits for and handles a WalletConnect session response with arbitrary data.
+    pub async fn send_session_request_and_wait<T, R, F>(
+        &self,
+        chain_id: &WcChainId,
+        method: WcRequestMethods,
+        params: serde_json::Value,
+        callback: F,
+    ) -> MmResult<R, WalletConnectError>
+    where
+        T: DeserializeOwned,
+        F: Fn(T) -> MmResult<R, WalletConnectError>,
+    {
+        let active_topic = self.session.get_active_topic_or_err().await?;
+        let request = SessionRequestRequest {
+            chain_id: chain_id.to_string(),
+            request: SessionRequest {
+                method: method.as_ref().to_string(),
+                expiry: Some(Utc::now().timestamp() as u64 + 60),
+                params,
+            },
+        };
+        let request = RequestParams::SessionRequest(request);
+        let ttl = request.irn_metadata().ttl;
+        self.publish_request(&active_topic, request).await?;
+
+        if let Ok(Some(resp)) = self.message_rx.lock().await.next().timeout_secs(ttl as f64).await {
```

Comment:
now after we publish our request, are we really sure that the next response we get from the message_rx was really intended for us?
what if there is another thread that sent a request at around the same time? how do we gaurantee each thread gets the resposne message it should be expecting?

Reply:
Yes, Since the wallet processes only one request at a time and responds in the same order (FIFO)

Reply:
[06e4dc7](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2223/commits/06e4dc7baeb40c8616cb3397fb3c44b91de191e6)